prov/verbs/src/verbs_domain_xrc.c

```diff
@@ -0,0 +1,576 @@
+{
+	struct fi_ibv_ep *ep;
+	enum fi_ibv_ini_qp_state last_state;
+	int ret;
+
+sched_next:
+	if (dlist_empty(&ini_conn->pending_list) ||
+			ini_conn->state == FI_IBV_INI_QP_CONNECTING)
+		return;
+
+	dlist_pop_front(&ini_conn->pending_list, struct fi_ibv_ep,
+			ep, ini_conn_entry);
+
+	last_state = ep->ini_conn->state;
+	if (last_state == FI_IBV_INI_QP_UNCONNECTED) {
+		ret = fi_ibv_create_ini_qp(ep);
+		if (ret) {
+			VERBS_WARN(FI_LOG_FABRIC,
+				   "Failed to create physical INI QP %d\n",
+				   ret);
+			abort();
+		}
+		ep->ini_conn->ini_qp = ep->id->qp;
+		ep->ini_conn->state = FI_IBV_INI_QP_CONNECTING;
+	} else {
+		if (!ep->id->qp) {
+			ep->conn_setup->rsvd_ini_qpn = fi_ibv_reserve_qpn(ep);
+			if (!ep->conn_setup->rsvd_ini_qpn) {
+				VERBS_WARN(FI_LOG_FABRIC,
+					  "rsvd_ini_qpn create err %d\n",
+					  errno);
+				abort();
+			}
+		}
+	}
+
+	assert(ep->ini_conn->ini_qp);
+
+	ep->ibv_qp = ep->ini_conn->ini_qp;
+	dlist_insert_tail(&ep->ini_conn_entry,
+			  &ep->ini_conn->active_list);
+
+	ret = fi_ibv_process_ini_conn(ep, ep->conn_setup->pending_recip,
+				      ep->conn_setup->pending_param,
+				      ep->conn_setup->pending_paramlen);
+	if (ret) {
+		ep->ini_conn->state = last_state;
+		fi_ibv_put_shared_ini_conn(ep);
+
+		/* We need to let the application know that the connect request
+		 * has failed. */
+		fi_ibv_create_shutdown_event(ep);
+	}
+
+	/* Continue to schedule shared connections if the physical connection
+	 * has completed and there are connection requests pending. We could
+	 * implement a throttle here if it is determined that it is better to
+	 * limit the number of outstanding connections. */
+	goto sched_next;
+
+	return;
+}
+
+/* Caller must hold domain:xrc:ini_mgmt_lock */
+int fi_ibv_process_ini_conn(struct fi_ibv_ep *ep,int reciprocal,
+			    void *param, size_t paramlen)
+{
+	struct fi_ibv_xrc_cm_data *cm_data = param;
+	struct rdma_conn_param conn_param = { 0 };
+	int ret;
+
+	assert(ep->ibv_qp);
+
+	if (!reciprocal)
+		fi_ibv_eq_set_xrc_conn_tag(ep);
+
+	fi_ibv_set_xrc_cm_data(cm_data, reciprocal, ep->conn_setup->tag,
+			       ep->eq->pep_port, ep->ini_conn->tgt_qpn);
+	conn_param.private_data = cm_data;
+	conn_param.private_data_len = paramlen;
+	conn_param.responder_resources = RDMA_MAX_RESP_RES;
+	conn_param.initiator_depth = RDMA_MAX_INIT_DEPTH;
+	conn_param.flow_control = 1;
+	conn_param.retry_count = 15;
+	conn_param.rnr_retry_count = 7;
+	conn_param.srq = 1;
+
+	/* Shared connections use reserved temporary QP numbers to
+	 * avoid the appearance of stale/duplicate CM messages */
+	if (!ep->id->qp)
+		conn_param.qp_num = ep->conn_setup->rsvd_ini_qpn->qp_num;
+
+	assert(ep->conn_state == FI_IBV_XRC_UNCONNECTED ||
+	       ep->conn_state == FI_IBV_XRC_ORIG_CONNECTED);
+	ep->conn_state++;
+
+	ret = rdma_connect(ep->id, &conn_param) ? -errno : 0;
+	if (ret) {
+		ret = -errno;
+		VERBS_INFO_ERRNO(FI_LOG_FABRIC, "rdma_connect", errno);
+		ep->conn_state--;
```

Comment:
It is very odd to see a variable called state decremented.  Is this a reference count or an actual state?

Reply:
It is rolling back the state after a failure.

Reply:
Can we replace the ++ / -- with state assignments?

Reply:
Sure. I have removed the ++/--.

The states are defined in an enum, and were intended to progress through sequentially as bidirectional XRC connections were created (and the previous state was validated with asserts), which is why the ++/-- was used.